### DISCUSSION TITLE
Least common denominator

### DISCUSSION THREAD

**Discussion by spamatica** [muse-sequencer/muse#1201] [Team Posts]
Trying to keep up with trends of our favourite tooling packages (well, not really trying, more like: sometimes stumbling on bits of information)
I saw qt6.0 was released the other month, at some point we should probably look at the portability. No hurry on my part 5.x will be supported for a few years still and 6.0 has probably not entered that many distributions yet.

This does bring me to my main thought though. Currently MusE supports:
Qt from version 5.7 (released 2016)
Gcc 4.x, or rather C++ 11 (4.8 with full c++11 was released in 2013)

For both of these I'm curious if there are systems out there (that are in our scope to support) that have such old libraries, or if we safely could increase the lowest version?

For Qt we have stumbled a few times, depending on things from newer versions but as I recall nothing major so we found other solutions, perhaps there isn't currently any need to increase this.

For C++ there has been three newer releases, 14, 17 and now 20. Not entirely sure what features they contain that we could make use of. Though I'm sure there are a bunch.

I was mostly interested if you had any thoughts on this? 
Going back to Qt6 I see it requires C++17.

**Comment by terminator356** [comment]
For Qt it might not be necessary to bump up the lowest version.
We already add conditional defines to support newer versions as they come along and obsolete things.
Bumping up the minimum would likely involve removing or 'cleaning' some of those older
 conditional defines we use. Would make it look prettier.
But then, what exact minimum would we set?

As for C++ yeah, I've been watching and reading and thinking a lot about it lately.
I would agree and look forward to 17, because it finally brings some needed things like
 being able to modify map keys **in-place** - no more removing then re-inserting items into the map.
C++ 20 is a completely different almost foreign-looking ballgame.
New concepts such as... 'concepts' (yeah that's what they're called).
It would sure take some time to get used to it.

So there are a few places where we already set (or force) the C++ version.
I think maybe in the top-level cmake and a few other places maybe plugins or LV2 module or something.
I expect that changing it to 17... might... possibly... reveal some required upgrades to existing code?

**Comment by kybos** [comment]
I think we should have some fix point as a reference. Many use the oldest supported LTS version of Ubuntu for that, which is probably the best thing you can get in the volatile Linux world.
Ubuntu 16.04 LTS reaches EOL in April 2021, after that we could safely adapt to the next one (18.04).
This means: 
gcc 7.5.0
Qt 5.9
And then perhaps also look into potential C++17 issues and make the switch (c++17 has been supported since gcc5).

I don't see much benefit in Qt6 ATM, there are hardly any essential improvements in the area of classic widgets AFAICJ.

**Comment by spamatica** [comment]
Sounds reasonable. At least until the imagined point where we really want to use a feature that requires a newer version.

**Comment by spamatica** [comment]
Out of curiosity I tried building with C++17 enabled and most of it went fine. Fails in scoreedit.cpp for std::map manipulation, didn't get exactly what was wrong so I commented out a bunch. The rest compiled.

**Comment by kybos** [comment]
I already have a branch waiting for Ubuntu 16.04 EOL with all the version updates and some cleanups. I'm also building that with c++17 but I don't get any errors in scoreedit.cpp, strange... I think there shouldn't be any incompatibilities in std::map.

**Comment by kybos** [comment]
The Ubuntu 16.04 EOL has been reached so I will merge the version updates soon (this weekend probably).
I've done some more testing but I can't see any problems with c++17 so I will update that too (it has a lot of nice new features).
I will also have to update the CI build scripts ("actions") to run on Ubuntu 18.04.

**Comment by terminator356** [comment]
Looking forward to the 17 features. Could really use them right now. Bring it on!

**Comment by kybos** [comment]
It's in master now. Let me know if there are any problems with the build on your side.

**Comment by terminator356** [comment]
Fails (so far) in score edit as you mentioned before, on two lines 2100 and 2591.
Lemme see if I can fix that...

**Comment by terminator356** [comment]
Fixed in master now.

**Comment by kybos** [comment]
Funny, no errors here, and the AppImage build works too.
Maybe you (and Robert) have a newer gcc version or something...

**Comment by kybos** [comment]
With your fix it compiles too here, so everything is good.
BTW: You can ignore the failing Compatibility build action for the time being, I will look into it further when I have time... It's not critical.

**Comment by terminator356** [comment]
I've not updated my rolling distro in about a year so you may actually have a newer version than me.
gcc --version: 
`gcc (SUSE Linux) 10.1.1 20200625`

**Comment by kybos** [comment]
Not really, I have 7.5.0 🙂
